### ISSUE TITLE
UWP Cache storage limit hit with b2clogin.com authority

### ISSUE THREAD

**Issue by aespino27** [AzureAD/microsoft-authentication-library-for-dotnet#737]
**Which Version of MSAL are you using ?**
MSAL 2.6.1

**Which platform has the issue?**
UWP

**What authentication flow has the issue?**
* Desktop
    * [ ] Interactive

Using MSAL 2.6.1 against Azure AD B2C.

As per [this](https://docs.microsoft.com/en-us/azure/active-directory-b2c/b2clogin) recommendation, I changed authority to b2clogin.com, instead of login.microsoftonline.com, and now application fails.

Yes, ValidateAuthority has been set to false, like this:
app = new PublicClientApplication(Constants.ClientId, Constants.Authority)
            {
                RedirectUri = Constants.MSALRedirect
                ,
                ValidateAuthority = false             //required if using b2clogin.com
            };

No, no Identity providers configured in Azure AD B2C, just native users.

In my Constants class
//public static string Authority = $"https://login.microsoftonline.com/tfp/{Tenant}/{PolicySignUpSignIn}/oauth2/v2.0/authorize";
became
public static string Authority = $"https://{myadb2ctenantshortname}.b2clogin.com/tfp/{Tenant}/{PolicySignUpSignIn}/oauth2/v2.0/authorize";
        

Error appears when calling AcquireTokenAsync, like this
try
{
    authResult = await app.AcquireTokenAsync(scopes);
    return authResult.AccessToken;
{

Exception is not even of type MSALsomething, had to be caught as System.Exception.
Here is the stack trace
   at System.Runtime.InteropServices.WindowsRuntime.IMap`2.Insert(K key, V value)
   at System.Runtime.InteropServices.WindowsRuntime.MapToDictionaryAdapter.Insert[K,V](IMap`2 _this, K key, V value)
   at System.Runtime.InteropServices.WindowsRuntime.MapToDictionaryAdapter.Indexer_Set[K,V](K key, V value)
   at Microsoft.Identity.Client.Platforms.uap.UapTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item)
   at Microsoft.Identity.Client.TelemetryCore.TelemetryTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item, RequestContext requestContext)
   at Microsoft.Identity.Client.TokenCache.SaveAccessAndRefreshToken(AuthenticationRequestParameters requestParams, MsalTokenResponse response)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.CacheTokenResponseAndCreateAuthenticationResult(MsalTokenResponse msalTokenResponse)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.PublicClientApplication.AcquireTokenForLoginHintCommonAsync(Authority authority, IEnumerable`1 scopes, IEnumerable`1 extraScopesToConsent, String loginHint, UIBehavior behavior, String extraQueryParameters, UIParent parent, ApiIds apiId)
   at Microsoft.Identity.Client.PublicClientApplication.AcquireTokenAsync(IEnumerable`1 scopes)

Here is the message:
Message = "The length of the state manager setting name has exceeded the limit.\r\n\r\nError trying to write application data container value"

Visual Studio debug output:
Exception thrown: 'System.Exception' in System.Private.CoreLib.dll
WinRT information: Error trying to write application data container value

**Expected behavior**
Be able to obtain AccessTokens from Azure AD B2C using b2clogin.com as I did when using login.microsoftonline.com

**Comment by jennyf19** [comment]
@aespino27 Was a previous version of MSAL working for you? if so, which one?

**Comment by aespino27** [comment]
I didn't try the b2clogin.com thing in any previous version of MSAL.
I've been using almost every single preview version up until today, as it seems it is finally GA.
So today I upgraded to 2.6.1 and was still using login.microsoftonline.com, but looking into the Azure AD B2C documentation I found the recommendation to move to b2clogin.com as login.microsoft.com will eventually be deprecated, and ran into the issue I reported.
As of now I switched back to use login.microsoftonline.com using MSAL 2.6.1 and everything is ok, but considering the b2clogin.com has been around since March, I supposed by this time MSAL should be able to handle it, but it seems not.

**Comment by jennyf19** [comment]
@aespino27 Msal has been able to handle b2clogin.com since [Msalv2.4.0](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/releases/tag/MSAL-v2.4.0), and we have UI automation tests and samples using the new authority. I think there's a different issue here. 

Things work fine with login.microsoftonline.com w/2.6.1? but not b2clogin.com?
Does b2clogin.com work with 2.6.0-preview?

Also, could  you send [logs](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/logging#sample-for-logging-in-msal-2x)? email works jeferrie@microsoft.com thanks.

**Comment by aespino27** [comment]
Q: Things work fine with login.microsoftonline.com w/2.6.1? but not b2clogin.com?
A: Yes

Q: Does b2clogin.com work with 2.6.0-preview?
A: No

Q: Also, could you send logs?
A: Showing logs using 2.6.0-preview, using 2.6.1 resulted in the exact same logs. As far as the logs "speak", when using b2clogin.com, it looks like it gets the token, but the problem is when trying to save it into the cache?

************* 	BEGIN USING LOGIN.MICROSOFTONLINE.COM	*************

Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:15] Deserialized 0 items to token cache.
Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
MsalUiRequiredException: Null account was passed in AcquiretokenSilent API. Pass in an account object or call acquireToken to authenticate.
Info (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:15 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] 
=== Request Data ===
Authority Provided? - True
Client Id - <my client id>
Scopes - https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
Redirect Uri - msal<my client id>://auth
Validate Authority? - True
LoginHint provided? - False
User provided? - False
Extra Query Params Keys (space separated) - 
Slice Parameters Keys(space separated) - 

Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:15 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Additional scopes - ;UIBehavior - select_account
Info (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:15 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] === Token Acquisition (InteractiveRequest) started:
	Authority: https://login.microsoftonline.com/tfp/<my tenant>.onmicrosoft.com/b2c_1_<my b2c policy>/
	Scope: https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
	ClientId: <my client id>
	Cache Provided: True
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files\WindowsApps\Microsoft.NET.CoreFramework.Debug.2.2_2.2.27031.1_x86__8wekyb3d8bbwe\System.Diagnostics.DiagnosticSource.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\IntelliTrace\Microsoft.IntelliTrace.TelemetryObserver.Common.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\IntelliTrace\Microsoft.IntelliTrace.TelemetryObserver.CoreClr.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:16 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Resolving authority endpoints... Already resolved? - False
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:16 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Is Authority <my tenant>less? - False
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Checking client info returned from the server..
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Saving Token Response to cache..
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Looking for scopes for the authority in the cache which intersect with https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Intersecting scope entries count - 0
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Matching entries after filtering by user - 0
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] Saving RT in cache...
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:04:36 - a79aeea0-4f6e-48c9-995f-1c2af34998b3] === Token Acquisition finished successfully. An access token was returned with Expiration Time: 12/19/2018 06:04:36 +00:00 ===


************* 	END USING LOGIN.MICROSOFTONLINE.COM	*************

************* 	BEGIN USING B2CLOGIN.COM	*************

Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42] Deserialized 0 items to token cache.
Info (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] 
=== Request Data ===
Authority Provided? - True
Client Id - <my client id>
Scopes - https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
Redirect Uri - msal<my client id>://auth
Validate Authority? - False
LoginHint provided? - False
User provided? - True
Extra Query Params Keys (space separated) - 
Slice Parameters Keys(space separated) - 

Info (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] === Token Acquisition (SilentRequest) started:
	Authority: https://<my tenant>.b2clogin.com/tfp/<my tenant>.onmicrosoft.com/b2c_1_<my b2c policy>/
	Scope: https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
	ClientId: <my client id>
	Cache Provided: True
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Looking up access token in the cache.
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Filtering by user identifier...
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Matching entry count -1
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Matching entry count after filtering by scopes - 1
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] No tokens found for matching authority, client_id, user and scopes.
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Looking up refresh token in the cache..
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Refresh token found in the cache? - False
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Checking ADAL cache for matching RT
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42] Deserialized 0 items to token cache.
Verbose (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] No Refresh Token was found in the cache
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files\WindowsApps\Microsoft.NET.CoreFramework.Debug.2.2_2.2.27031.1_x86__8wekyb3d8bbwe\System.Diagnostics.StackTrace.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files\WindowsApps\Microsoft.NET.CoreFramework.Debug.2.2_2.2.27031.1_x86__8wekyb3d8bbwe\System.Reflection.Metadata.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files\WindowsApps\Microsoft.NET.CoreFramework.Debug.2.2_2.2.27031.1_x86__8wekyb3d8bbwe\System.IO.FileSystem.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files\WindowsApps\Microsoft.NET.CoreFramework.Debug.2.2_2.2.27031.1_x86__8wekyb3d8bbwe\System.Collections.Immutable.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Error (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 178156ed-5cfd-4a06-85a9-2b834ddf5069] Microsoft.Identity.Client.MsalUiRequiredException: No Refresh Token found in the cache
   at Microsoft.Identity.Client.Internal.Requests.SilentRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
	ErrorCode: no_tokens_found
	StatusCode: 0
	Claims: 
Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
MsalUiRequiredException: No Refresh Token found in the cache
Info (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 3ef00713-8997-4946-bc54-5de769b9f47c] 
=== Request Data ===
Authority Provided? - True
Client Id - <my client id>
Scopes - https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
Redirect Uri - msal<my client id>://auth
Validate Authority? - False
LoginHint provided? - False
User provided? - False
Extra Query Params Keys (space separated) - 
Slice Parameters Keys(space separated) - 

Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 3ef00713-8997-4946-bc54-5de769b9f47c] Additional scopes - ;UIBehavior - select_account
Info (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 3ef00713-8997-4946-bc54-5de769b9f47c] === Token Acquisition (InteractiveRequest) started:
	Authority: https://<my tenant>.b2clogin.com/tfp/<my tenant>.onmicrosoft.com/b2c_1_<my b2c policy>/
	Scope: https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
	ClientId: <my client id>
	Cache Provided: True
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 3ef00713-8997-4946-bc54-5de769b9f47c] Resolving authority endpoints... Already resolved? - False
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:10:42 - 3ef00713-8997-4946-bc54-5de769b9f47c] Is Authority <my tenant>less? - False
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files\WindowsApps\Microsoft.NET.CoreFramework.Debug.2.2_2.2.27031.1_x86__8wekyb3d8bbwe\System.Diagnostics.DiagnosticSource.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\IntelliTrace\Microsoft.IntelliTrace.TelemetryObserver.Common.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'Domino.Client.UWP.exe' (CoreCLR: CoreCLR_UWP_Domain): Loaded 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\IntelliTrace\Microsoft.IntelliTrace.TelemetryObserver.CoreClr.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:11:05 - 3ef00713-8997-4946-bc54-5de769b9f47c] Checking client info returned from the server..
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:11:05 - 3ef00713-8997-4946-bc54-5de769b9f47c] Saving Token Response to cache..
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:11:05 - 3ef00713-8997-4946-bc54-5de769b9f47c] Looking for scopes for the authority in the cache which intersect with https://<my tenant>.onmicrosoft.com/<my app id uri>/<my scope>
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:11:05 - 3ef00713-8997-4946-bc54-5de769b9f47c] Intersecting scope entries count - 0
Info (False) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:11:05 - 3ef00713-8997-4946-bc54-5de769b9f47c] Matching entries after filtering by user - 0
Error (True) MSAL 2.6.0.0 MSAL.UAP N/A [12/19/2018 05:11:05 - 3ef00713-8997-4946-bc54-5de769b9f47c] System.Exception: The length of the state manager setting name has exceeded the limit.

Error trying to write application data container value
   at System.Runtime.InteropServices.WindowsRuntime.IMap`2.Insert(K key, V value)
   at System.Runtime.InteropServices.WindowsRuntime.MapToDictionaryAdapter.Insert[K,V](IMap`2 _this, K key, V value)
   at System.Runtime.InteropServices.WindowsRuntime.MapToDictionaryAdapter.Indexer_Set[K,V](K key, V value)
   at Microsoft.Identity.Client.Platforms.uap.UapTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item)
   at Microsoft.Identity.Client.TelemetryCore.TelemetryTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item, RequestContext requestContext)
   at Microsoft.Identity.Client.TokenCache.SaveAccessAndRefreshToken(AuthenticationRequestParameters requestParams, MsalTokenResponse response)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.CacheTokenResponseAndCreateAuthenticationResult(MsalTokenResponse msalTokenResponse)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
Exception thrown: 'System.Exception' in System.Private.CoreLib.dll
WinRT information: Error trying to write application data container value


************* 	END USING B2CLOGIN.COM		*************

**Comment by jennyf19** [comment]
Thank you @aespino27

**Comment by jmprieur** [comment]
@aespino27 : the authority should not have the `oauth2/v2.0/authorize` part (what you have described is an endpoint, not an authority).
Did you try with `$"https://{myadb2ctenantshortname}.b2clogin.com/tfp/{Tenant}/{PolicySignUpSignIn}/"` ?

**Comment by aespino27** [comment]
@jmprieur : Removing the oauth2/v2.0/authorize, that is, using $"https://{myadb2ctenantshortname}.b2clogin.com/tfp/{Tenant}/{PolicySignUpSignIn}/", yields to the exact same exception and stack trace … System.Exception: The length of the state manager setting name has exceeded the limit.

So @jennyf19, I concur it might be related to hitting the cache storage limit for UWP, I mean, the stack trace blames Microsoft.Identity.Client.Platforms.uap.UapTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item), which is shown below from the source code

public void SaveAccessToken(MsalAccessTokenCacheItem item)
{
    ApplicationDataCompositeValue composite = new ApplicationDataCompositeValue();
    SetCacheValue(composite, JsonHelper.SerializeToJson(item));
    var key = item.GetKey().GetUWPFixedSizeKey();
    _accessTokenContainer.Values[/*CoreCryptographyHelpers.CreateBase64UrlEncodedSha256Hash(cacheKey)*/key] = composite;
}

It looks like either the key (key) or the value (composite) length is hitting some limit.
the previous piece of code calls, GetUWPFixedSizeKey, which in turn calls MsalCacheCommon.GetCredentialKey

#region UWP
    /// <summary>
    /// Gets a key that is smaller than 255 characters, which is a limitation for 
    /// UWP storage. This is done by hashing the scopes and env.
    /// </summary>
    /// <remarks>
    /// accountId - two guids plus separator - 73 chars        
    /// "accesstoken" string - 11 chars
    /// env - ussually loging.microsoft.net - 20 chars
    /// clientid - a guid - 36 chars
    /// tenantid - a guid - 36 chars
    /// scopes - a sha256 string - 44 chars
    /// delimiters - 4 chars
    /// total: 224 chars
    /// </remarks>
public string GetUWPFixedSizeKey()
{
    var crypto = PlatformProxyFactory.GetPlatformProxy().CryptographyManager;
    return MsalCacheCommon.GetCredentialKey(
        _homeAccountId,
        _environment,
        MsalCacheCommon.AccessToken,
        _clientId,
        _tenantId,
        crypto.CreateSha256Hash(_normalizedScopes)); // can't use scopes and env because they are of variable length
}
#endregion

But I don't see in MsalCacheCommon.GetCredentialKey any validation if the returning string is greater than 255

**Comment by bgavrilMS** [comment]
@aespino27 - how many does $myadb2ctenantshortname have?

**Comment by aespino27** [comment]
23

**Comment by bgavrilMS** [comment]
Understood, the url is long and is causing the issue. I will look at providing a fix as soon as possible.

**Comment by aespino27** [comment]
Thx

**Comment by bgavrilMS** [comment]
Hi @aespino27 - apologies for the long wait. I have a fix for this, would it be possible for you to try it out? 

I created a nuget packge with MSAL with the fix and you can get it here: https://www.myget.org/feed/aad-clients-nightly/package/nuget/Microsoft.Identity.Client/2.7.1-internal02302uwp2x

If you don't know how to install nuget package from sources other than NuGet.org, check this wiki page out - https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Installing-a-nuget-package-from-a-source-other-than-NuGet.org

**Comment by aespino27** [comment]
Thank you, I'll try to give it a shot during the weekend and let you know the outcome.

**Comment by MattLisson** [comment]
I'm still having this issue even with your fixed build: Microsoft.Identity.Client/2.7.1-internal02302uwp2x

I've altered the UWP test app in this Repo to try and find out why I'm getting a longer Key, and found one assumption in ```GetUWPFixedSizedKey()``` that doesn't hold:
 'accountId' is two guids plus separator (73 chars).

My ```MsalTokenResponse``` from ```InteractiveRequest.SendTokenRequestAsync``` contains ClientInfo that decodes to this (Guids created randomly to demonstrate):
```
{
  "uid":"db7f18b9-e336-4dfe-adef-c23d321103fb-b2c_1_signupsignin",
  "utid":"95ff1110-5b50-400a-898e-515edf66e84a"
}
```
Notice that the policy has been tacked onto the ```uid``` by the server. Looking in Azure Portal, the Guid part of uid is correct for the user, so I'm not sure why the policy was appended. The extra characters of the policy is enough to push the string length above 255, since the ClientInfo is used in full within the cache key.

_PS: The ```GetUWPFixedSizeKey``` assumption: ```delimiters - 4 chars```, is also incorrect, there are 5 delimiters in ```MsalCacheCommon.GetCredentialKey```._

**Comment by bgavrilMS** [comment]
Thanks @MattLisson , much appreciated. I can hash the entire key to ensure it is used as is. 

Alternatively, I can move the token cache from the "native" platform dictionary to a blob, which will need to be loaded everytime. Trying to avoid this as it means that we'd have to either migrate the token cache or break existing users (i.e. when you upgrade your UWP app, your users will have to re-sign in).

**Comment by bgavrilMS** [comment]
The 3.0.0-preview release will have a fix for this. It should be out in 1-2 weeks max.

**Comment by jmprieur** [comment]
@aespino27 : the MSAL.NET 3.0.0-preview is out. See https://aka.ms/msal-net-3x
cc: @bgavrilMS

**Comment by jennyf19** [comment]
Fixed in [MSAL 3.0.0-preview release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/3.0.0-preview)
